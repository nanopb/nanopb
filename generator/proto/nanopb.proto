// This file contains definitions of custom options used to control the
// code generator in nanopb protocol buffers library.
//
// Most commonly used options are max_count and max_size, which allow
// the generator to allocate static arrays for repeated and string fields.
//
// There are three ways to use these options:
//  1. Use a separate <protofile>.options file
//  2. Use command line switches to nanopb_generator.py
//  3. Use [(nanopb).option = value] in your <protofile>.proto file
//
// For detailed documentation, refer to "Generator options" in docs/reference.md

syntax = "proto2";
import "google/protobuf/descriptor.proto";

option java_package = "fi.kapsi.koti.jpa.nanopb";

enum FieldType {
    FT_DEFAULT = 0; // Automatically decide field type, generate static field if possible.
    FT_CALLBACK = 1; // Always generate a callback field.
    FT_POINTER = 4; // Always generate a dynamically allocated field.
    FT_STATIC = 2; // Generate a static field or raise an exception if not possible.
    FT_IGNORE = 3; // Ignore the field completely.
    FT_INLINE = 5; // Legacy option, use the separate 'fixed_length' option instead
}

enum IntSize {
    IS_DEFAULT = 0; // Default, 32/64bit based on type in .proto
    IS_8 = 8;
    IS_16 = 16;
    IS_32 = 32;
    IS_64 = 64;
}

enum TypenameMangling {
    M_NONE = 0; // Default, no typename mangling
    M_STRIP_PACKAGE = 1; // Strip current package name
    M_FLATTEN = 2; // Only use last path component
    M_PACKAGE_INITIALS = 3; // Replace the package name by the initials
}

enum DescriptorSize {
    DS_AUTO = 0; // Select minimal size based on field type
    DS_1 = 1;    // 1 word; up to 15 byte fields, no arrays
    DS_2 = 2;    // 2 words; up to 4095 byte fields, 4095 entry arrays
    DS_4 = 4;    // 4 words; up to 2^32-1 byte fields, 2^16-1 entry arrays
    DS_8 = 8;    // 8 words; up to 2^32-1 entry arrays
}

// This is the inner options message, which basically defines options for
// a field. When it is used in message or file scope, it applies to all
// fields.
message NanoPBOptions {
  // Allocated size for 'bytes' and 'string' fields.
  // For string fields, this should include the space for null terminator.
  optional int32 max_size = 1;
  
  // Maximum length for 'string' fields. Setting this is equivalent
  // to setting max_size to a value of length+1.
  optional int32 max_length = 14;
  
  // Allocated number of entries in arrays ('repeated' fields)
  optional int32 max_count = 2;
  
  // Size of integer fields. Can save some memory if you don't need
  // full 32 bits for the value.
  optional IntSize int_size = 7 [default = IS_DEFAULT];

  // Size for enum fields. Supported by C++11 and C23 standards.
  optional IntSize enum_intsize = 34 [default = IS_DEFAULT];

  // Force type of field (callback or static allocation)
  optional FieldType type = 3 [default = FT_DEFAULT];
  
  // Use long names for enums, i.e. EnumName_EnumValue.
  optional bool long_names = 4 [default = true];
  
  // Add 'packed' attribute to generated structs.
  // Note: this cannot be used on CPUs that break on unaligned
  // accesses to variables.
  optional bool packed_struct = 5 [default = false];
  
  // Add 'packed' attribute to generated enums.
  optional bool packed_enum = 10 [default = false];
  
  // Skip this message
  optional bool skip_message = 6 [default = false];

  // Generate oneof fields as normal optional fields instead of union.
  optional bool no_unions = 8 [default = false];

  // integer type tag for a message
  optional uint32 msgid = 9;

  // decode oneof as anonymous union
  optional bool anonymous_oneof = 11 [default = false];

  // Proto3 singular field does not generate a "has_" flag
  optional bool proto3 = 12 [default = false];
  
  // Force proto3 messages to have no "has_" flag.
  // This was default behavior until nanopb-0.4.0.
  optional bool proto3_singular_msgs = 21 [default = false];

  // Generate an enum->string mapping function (can take up lots of space).
  optional bool enum_to_string = 13 [default = false];

  // Generate validation methods for enums
  optional bool enum_validate = 32 [default = false];

  // Generate bytes arrays with fixed length
  optional bool fixed_length = 15 [default = false];

  // Generate repeated field with fixed count
  optional bool fixed_count = 16 [default = false];

  // Generate message-level callback that is called before decoding submessages.
  // This can be used to set callback fields for submsgs inside oneofs.
  optional bool submsg_callback = 22 [default = false];

  // Shorten or remove package names from type names.
  // This option applies only on the file level.
  optional TypenameMangling mangle_names = 17 [default = M_NONE];

  // Data type for storage associated with callback fields.
  optional string callback_datatype = 18 [default = "pb_callback_t"];

  // Callback function used for encoding and decoding.
  // Prior to nanopb-0.4.0, the callback was specified in per-field pb_callback_t
  // structure. This is still supported, but does not work inside e.g. oneof or pointer
  // fields. Instead, a new method allows specifying a per-message callback that
  // will be called for all callback fields in a message type.
  optional string callback_function = 19 [default = "pb_default_field_callback"];

  // Select the size of field descriptors. This option has to be defined
  // for the whole message, not per-field. Usually automatic selection is
  // ok, but if it results in compilation errors you can increase the field
  // size here.
  optional DescriptorSize descriptorsize = 20 [default = DS_AUTO];

  // Set default value for has_ fields.
  optional bool default_has = 23 [default = false];

  // Extra files to include in generated `.pb.h`
  repeated string include = 24;

  // Automatic includes to exclude from generated `.pb.h`
  // Same as nanopb_generator.py command line flag -x.
  repeated string exclude = 26;

  // Package name that applies only for nanopb.
  optional string package = 25;
  
  // Override type of the field in generated C code. Only to be used with related field types
  optional google.protobuf.FieldDescriptorProto.Type type_override = 27;

  // Override of the label of the field (see FieldDescriptorProto.Label). Can be used to create
  // fields which nanopb considers required in proto3, or whether nanopb treats the field as
  // optional/required/repeated.
  optional google.protobuf.FieldDescriptorProto.Label label_override = 31;

  // Due to historical reasons, nanopb orders fields in structs by their tag number
  // instead of the order in .proto. Set this to false to keep the .proto order.
  // The default value will probably change to false in nanopb-0.5.0.
  optional bool sort_by_tag = 28 [default = true];

  // Set the FT_DEFAULT field conversion strategy.
  // A field that can become a static member of a c struct (e.g. int, bool, etc)
  // will be a a static field.
  // Fields with dynamic length are converted to either a pointer or a callback.
  optional FieldType fallback_type = 29 [default = FT_CALLBACK]; 

  // Override initializer used in generated MyMessage_init_zero and MyMessage_init_default macros
  // By default decided automatically based on field default value and datatype.
  optional string initializer = 30;

  // Discard unused types that are automatically generated by protoc if they are not actually
  // needed. Currently this applies to map< > types when the field is ignored by options.
  optional bool discard_unused_automatic_types = 33 [default = true];

  // Discard messages and fields marked with [deprecated = true] in the proto file.
  optional bool discard_deprecated = 35 [default = false];
}

// Extensions to protoc 'Descriptor' type in order to define options
// inside a .proto file.
//
// Protocol Buffers extension number registry
// --------------------------------
// Project:  Nanopb
// Contact:  Petteri Aimonen <jpa@kapsi.fi>
// Web site: http://kapsi.fi/~jpa/nanopb
// Extensions: 1010 (all types)
// --------------------------------

extend google.protobuf.FileOptions {
    optional NanoPBOptions nanopb_fileopt = 1010;
}

extend google.protobuf.MessageOptions {
    optional NanoPBOptions nanopb_msgopt = 1010;
}

extend google.protobuf.EnumOptions {
    optional NanoPBOptions nanopb_enumopt = 1010;
}

extend google.protobuf.FieldOptions {
    optional NanoPBOptions nanopb = 1010;
}


