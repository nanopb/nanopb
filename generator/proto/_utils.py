import sys
import subprocess
import os.path

import traceback

try:
    # For PyInstaller compatibility with grpcio-tools >= 1.69
    import grpc_tools.grpc_version
except ImportError:
    pass

def has_grpcio_protoc(verbose = False):
    # type: () -> bool
    """ checks if grpcio-tools protoc is installed"""

    try:
        import grpc_tools.protoc
    except ImportError:
        if verbose:
            sys.stderr.write("Failed to import grpc_tools: %s\n" % traceback.format_exc())
        return False

    return True

def get_grpc_tools_proto_path():
    if sys.hexversion > 0x03090000:
        import importlib.resources as ir
        with ir.as_file(ir.files('grpc_tools') / '_proto') as path:
            return str(path)
    else:
        import pkg_resources
        return pkg_resources.resource_filename('grpc_tools', '_proto')

def get_proto_builtin_include_path():
    """Find include path for standard google/protobuf includes and for
    nanopb.proto.
    """

    if getattr(sys, 'frozen', False):
        # Pyinstaller package
        paths = [
            os.path.join(os.path.dirname(os.path.abspath(sys.executable)), 'proto'),
            os.path.join(os.path.dirname(os.path.abspath(sys.executable)), 'grpc_tools', '_proto')
        ]

    else:
        # Stand-alone script
        paths = [
            os.path.dirname(os.path.abspath(__file__))
        ]

        if has_grpcio_protoc():
            paths.append(get_grpc_tools_proto_path())

    return paths

def invoke_protoc(argv):
    # type: (list) -> typing.Any
    """
    Invoke protoc.

    This routine will use grpcio-provided protoc if it exists,
    using system-installed protoc as a fallback.

    Args:
        argv: protoc CLI invocation, first item must be 'protoc'
    """

    # Add current directory to include path if nothing else is specified
    if not [x for x in argv if x.startswith('-I')]:
        argv.append("-I.")

    # Add default protoc include paths
    for incpath in get_proto_builtin_include_path():
        argv.append('-I' + incpath)

    if has_grpcio_protoc():
        import grpc_tools.protoc as protoc
        return protoc.main(argv)
    else:
        return subprocess.call(argv)

def print_versions():
    try:
        if has_grpcio_protoc(verbose = True):
            import grpc_tools.protoc
            sys.stderr.write("Using grpcio-tools protoc from " + grpc_tools.protoc.__file__ + "\n")
        else:
            sys.stderr.write("Using protoc from system path\n")

        invoke_protoc(['protoc', '--version'])
    except Exception as e:
        sys.stderr.write("Failed to determine protoc version: " + str(e) + "\n")

    try:
        sys.stderr.write("protoc builtin include path: " + str(get_proto_builtin_include_path()) + "\n")
    except Exception as e:
        sys.stderr.write("Failed to construct protoc include path: " + str(e) + "\n")

    try:
        import google.protobuf
        sys.stderr.write("Python version " + sys.version + "\n")
        sys.stderr.write("Using python-protobuf from " + google.protobuf.__file__ + "\n")
        sys.stderr.write("Python-protobuf version: " + google.protobuf.__version__ + "\n")
    except Exception as e:
        sys.stderr.write("Failed to determine python-protobuf version: " + str(e) + "\n")

if __name__ == '__main__':
    print_versions()
