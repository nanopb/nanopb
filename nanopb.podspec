Pod::Spec.new do |s|
  s.name         = "nanopb"
  # CocoaPods minor version is minor * 10,000 + patch * 100 + fourth
  s.version      = "3.30910.0"
  s.summary      = "Protocol buffers with small code size."

  s.description  = <<-DESC
                    Nanopb is a small code-size Protocol Buffers implementation
                    in ansi C. It is especially suitable for use in
                    microcontrollers, but fits any memory restricted system.
                   DESC

  s.homepage     = "https://github.com/nanopb/nanopb"
  s.license      = { :type => 'zlib', :file => 'LICENSE.txt' }
  s.author       = { "Petteri Aimonen" => "jpa@nanopb.mail.kapsi.fi" }
  s.source       = { :git => "https://github.com/nanopb/nanopb.git", :tag => "0.3.9.10" }

  s.ios.deployment_target = '12.0'
  s.osx.deployment_target = '10.15'
  s.tvos.deployment_target = '13.0'
  s.watchos.deployment_target = '7.0'

  s.requires_arc = false
  s.xcconfig = { 'GCC_PREPROCESSOR_DEFINITIONS' => '$(inherited) PB_FIELD_32BIT=1 PB_NO_PACKED_STRUCTS=1 PB_ENABLE_MALLOC=1' }

  s.source_files  = 'nanopb/*.{h,c}'
  s.public_header_files = 'nanopb/*.h'

  s.subspec 'encode' do |e|
    e.public_header_files = ['pb.h', 'pb_encode.h', 'pb_common.h']
    e.source_files = ['pb.h', 'pb_common.h', 'pb_common.c', 'pb_encode.h', 'pb_encode.c']
  end

  s.subspec 'decode' do |d|
    d.public_header_files = ['pb.h', 'pb_decode.h', 'pb_common.h']
    d.source_files = ['pb.h', 'pb_common.h', 'pb_common.c', 'pb_decode.h', 'pb_decode.c']
  end

  s.module_name  = "nanopb"

  s.resource_bundles = {
    "#{s.module_name}_Privacy" => 'spm_resources/PrivacyInfo.xcprivacy'
  }
end
