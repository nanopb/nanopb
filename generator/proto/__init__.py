'''This file dynamically builds the proto definitions for Python.'''
from __future__ import absolute_import

import os
import os.path
import sys
import tempfile
import shutil
import traceback
from tempfile import TemporaryDirectory
from ._utils import has_grpcio_protoc, invoke_protoc, print_versions

def build_nanopb_proto(protosrc, dirname):
    '''Try to build a .proto file for python-protobuf.
    Returns True if successful.
    '''

    cmd = [
        "protoc",
        "--python_out={}".format(dirname),
        protosrc,
        "-I={}".format(dirname),
    ]

    if has_grpcio_protoc():
        # grpcio-tools has an extra CLI argument
        # from grpc.tools.protoc __main__ invocation.
        cmd.append("-I={}".format(_utils.get_grpc_tools_proto_path()))

    try:
        invoke_protoc(argv=cmd)
    except:
        sys.stderr.write("Failed to build nanopb_pb2.py: " + ' '.join(cmd) + "\n")
        sys.stderr.write(traceback.format_exc() + "\n")
        return False

    return True

def load_nanopb_pb2():
    # To work, the generator needs python-protobuf built version of nanopb.proto.
    # There are three methods to provide this:
    #
    # 1) Load a previously generated generator/proto/nanopb_pb2.py
    # 2) Use protoc to build it and store it permanently generator/proto/nanopb_pb2.py
    # 3) Use protoc to build it, but store only temporarily in system-wide temp folder
    #
    # By default these are tried in numeric order.
    # If NANOPB_PB2_TEMP_DIR environment variable is defined, the 2) is skipped.
    # If the value of the $NANOPB_PB2_TEMP_DIR exists as a directory, it is used instead
    # of system temp folder.

    tmpdir = os.getenv("NANOPB_PB2_TEMP_DIR")
    temporary_only = (tmpdir is not None)
    dirname = os.path.dirname(__file__)
    protosrc = os.path.join(dirname, "nanopb.proto")
    protodst = os.path.join(dirname, "nanopb_pb2.py")

    if tmpdir is not None and not os.path.isdir(tmpdir):
        tmpdir = None # Use system-wide temp dir

    no_rebuild = bool(int(os.getenv("NANOPB_PB2_NO_REBUILD", default = 0)))
    if bool(no_rebuild):
        # Don't attempt to autogenerate nanopb_pb2.py, external build rules
        # should have already done so.
        import nanopb_pb2 as nanopb_pb2_mod
        return nanopb_pb2_mod

    if os.path.isfile(protosrc):
        src_date = os.path.getmtime(protosrc)
        if os.path.isfile(protodst) and os.path.getmtime(protodst) >= src_date:
            try:
                from . import nanopb_pb2 as nanopb_pb2_mod
                return nanopb_pb2_mod
            except Exception as e:
                sys.stderr.write("Failed to import nanopb_pb2.py: " + str(e) + "\n"
                                "Will automatically attempt to rebuild this.\n"
                                "Verify that python-protobuf and protoc versions match.\n")
                print_versions()

    # Try to rebuild into generator/proto directory
    if not temporary_only:
        build_nanopb_proto(protosrc, dirname)

        try:
            from . import nanopb_pb2 as nanopb_pb2_mod
            return nanopb_pb2_mod
        except:
            sys.stderr.write("Failed to import generator/proto/nanopb_pb2.py:\n")
            sys.stderr.write(traceback.format_exc() + "\n")

    # Try to rebuild into temporary directory
    with TemporaryDirectory(prefix = 'nanopb-', dir = tmpdir) as protodir:
        build_nanopb_proto(protosrc, protodir)

        if protodir not in sys.path:
            sys.path.insert(0, protodir)

        try:
            import nanopb_pb2 as nanopb_pb2_mod
            return nanopb_pb2_mod
        except:
            sys.stderr.write("Failed to import %s/nanopb_pb2.py:\n" % protodir)
            sys.stderr.write(traceback.format_exc() + "\n")

    # If everything fails
    sys.stderr.write("\n\nGenerating nanopb_pb2.py failed.\n")
    sys.stderr.write("Make sure that a protoc generator is available and matches python-protobuf version.\n")
    print_versions()
    sys.exit(1)

