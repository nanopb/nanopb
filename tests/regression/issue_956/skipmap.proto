syntax = "proto3";
import "nanopb.proto";

option (nanopb_fileopt).max_count = 4;

message Message1 {
    map<uint32, string> mymap1 = 1;
}

message Message2 {
    option (nanopb_msgopt).skip_message = true;
    map<uint32, string> mymap2 = 1;
}

message Message3 {
    map<uint32, string> mymap3 = 1 [(nanopb).type = FT_IGNORE];
}

message Message4 {
    option (nanopb_msgopt).skip_message = true;

    message SubMessage4 {
        option (nanopb_msgopt).skip_message = true;
        uint32 dummy = 1;
    }

    map<uint32, SubMessage4> mymap4 = 1;
}
