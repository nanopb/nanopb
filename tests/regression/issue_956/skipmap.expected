Message1_Mymap1Entry mymap1
! Message2_Mymap2Entry
! Message3_Mymap3Entry
! Message4_Mymap4Entry
